torch/distributed/_mesh_layout.py

```diff
@@ -62,14 +76,288 @@ def __getitem__(self, i: int) -> "_Layout":
         else:
             raise ValueError("size and stride must be either int or tuple")
 
+    @staticmethod
+    def ceil_div(n: int, m: int) -> int:
+        return (n + m - 1) // m
+
     def coalesce(self) -> "_Layout":
-        layout = coalesce(self)
-        return _Layout(layout.shape, layout.stride)  # type: ignore[arg-type]
+        """
+        A layout is represented by (sizes):(strides), e.g. (3,2):(4,2).
+        Two consecutive dimensions can be "merged" into one if their
+        strides are contiguous/multiplicative (i.e., the inner stride * inner size
+        equals the next stride), we perform this kind of merge inside coalesce.
+
+        Example 1 (simple): (3,2):(2,1)
+        - inner dimension: has stride=1, size=2
+        - outer dimension: stride = inner_stride * inner_size = 2
+        → coalesced = (6:1)    # acts like a flat 1D array of length 6
+
+        Example 2 (non-coalescible): (3,2):(4,1)
+        - inner dimension: stride=1, size=2 → 2*1 = 2
+        - outer dimension: stride=4, mismatch (≠ 2)
+        → cannot merge; result stays (3,2):(4,1)
+        """
+        # Option 1: leverage pycute's coalesce
```

Comment:
There are two options: either we leverage existing coalesce from pycute but we somehow need to convert from lexicographic to co-lexicographic or we directly implement it inside `coalesce`. I personally like option2 but option1 is safer because we are leverage existing logic. Some things for `complement and composition`.

@ezyang Any comments on this?

Reply:
IIUC, we only need to change coalesce and complement. The composition part is actually fine, it works for both co-lexicographic and lexicographic.

Reply:
So I had originally intended for the change to be done in the pycute implementation directly. But it is true that changing it in a wrapper is simple and will obviously work. Hmm.

One benefit to changing it in the pycute implementation is you get test coverage automatically. If you change it here you have to port the tests to also operate on the wrapper class. I also am not a huge fan of the wrapper class but ehhhh.